src/services/codefixes/inferFromUsage.ts

```diff
@@ -0,0 +1,590 @@
+
+            // Property declarations
+            Diagnostics.Member_0_implicitly_has_an_1_type.code,
+        ],
+        getCodeActions: getActionsForAddExplicitTypeAnnotation
+    });
+
+    function getActionsForAddExplicitTypeAnnotation({ sourceFile, program, span: { start }, errorCode, cancellationToken }: CodeFixContext): CodeAction[] | undefined {
+        const token = getTokenAtPosition(sourceFile, start);
+
+        switch (token.kind) {
+            case SyntaxKind.Identifier:
+            case SyntaxKind.DotDotDotToken:
+            case SyntaxKind.PublicKeyword:
+            case SyntaxKind.PrivateKeyword:
+            case SyntaxKind.ProtectedKeyword:
+            case SyntaxKind.ReadonlyKeyword:
+                // Allowed
+                break;
+            default:
+                return undefined;
+        }
+
+        const containingFunction = getContainingFunction(token);
+        const checker = program.getTypeChecker();
+
+        switch (errorCode) {
+            // Variable and Property declarations
+            case Diagnostics.Member_0_implicitly_has_an_1_type.code:
+            case Diagnostics.Variable_0_implicitly_has_type_1_in_some_locations_where_its_type_cannot_be_determined.code:
+                return getCodeActionForVariableDeclaration(<PropertyDeclaration | PropertySignature | VariableDeclaration>token.parent);
+            case Diagnostics.Variable_0_implicitly_has_an_1_type.code:
+                return getCodeActionForVariableUsage(<Identifier>token);
+
+            // Paramtert declarations
+            case Diagnostics.Parameter_0_implicitly_has_an_1_type.code:
+                if (isSetAccessor(containingFunction)) {
+                    return getCodeActionForSetAccessor(containingFunction);
+                }
+            // falls through
+            case Diagnostics.Rest_parameter_0_implicitly_has_an_any_type.code:
+                return getCodeActionForParameter(<ParameterDeclaration>token.parent);
+
+            // Get Accessor declarations
+            case Diagnostics.Property_0_implicitly_has_type_any_because_its_get_accessor_lacks_a_return_type_annotation.code:
+            case Diagnostics._0_which_lacks_return_type_annotation_implicitly_has_an_1_return_type.code:
+                return isGetAccessor(containingFunction) ? getCodeActionForGetAccessor(containingFunction) : undefined;
+
+            // Set Accessor declarations
+            case Diagnostics.Property_0_implicitly_has_type_any_because_its_set_accessor_lacks_a_parameter_type_annotation.code:
+                return isSetAccessor(containingFunction) ? getCodeActionForSetAccessor(containingFunction) : undefined;
+
+            // Property declarations
+            case Diagnostics.Member_0_implicitly_has_an_1_type.code:
+                return getCodeActionForVariableDeclaration(<VariableDeclaration>token.parent);

```

Comment:
Is there ever a time that you won't have a symbol that you can't call this directly? What about in the case of a computed property?

Reply:
We filter these earlier on. it can only be identifier for a var declarations